weaviate/connect/connection.py

```diff
@@ -190,21 +190,28 @@ def _create_background_token_refresh(self, _auth: Optional[_Auth] = None):
         self._shutdown_background_event = Event()
 
         def periodic_refresh_token(refresh_time: int, _auth: Optional[_Auth]):
-            time.sleep(max(refresh_time - 5, 1))
+            time.sleep(max(refresh_time - 30, 1))
             while not self._shutdown_background_event.is_set():
                 # use refresh token when available
-                if "refresh_token" in self._session.token:
-                    self._session.token = self._session.refresh_token(
-                        self._session.metadata["token_endpoint"]
-                    )
-                    refresh_time = self._session.token.get("expires_in")
-                else:
-                    # client credentials usually does not contain a refresh token => get a new token using the
-                    # saved credentials
-                    assert _auth is not None
-                    new_session = _auth.get_auth_session()
-                    self._session.token = new_session.fetch_token()
-                time.sleep(max(refresh_time - 5, 1))
+                try:
+                    if "refresh_token" in self._session.token:
+                        assert isinstance(self._session, OAuth2Session)
+                        self._session.token = self._session.refresh_token(
+                            self._session.metadata["token_endpoint"]
+                        )
+                        refresh_time = self._session.token.get("expires_in") - 30
+                    else:
+                        # client credentials usually does not contain a refresh token => get a new token using the
+                        # saved credentials
+                        assert _auth is not None
+                        new_session = _auth.get_auth_session()
+                        self._session.token = new_session.fetch_token()
+                except (RequestsHTTPError, ReadTimeout) as exc:
+                    # retry again after one second, might be an unstable connection
+                    refresh_time = 1
+                    _Warnings.token_refresh_failed(exc)
```

Comment:
I think it would be better to use `logging` here instead of the `warnings` library. The `warnings` library is generally intended for situations that could have been provided if the client code would have been adjusted. That is not the case here. The challenge is that in a production setting usually your monitoring tooling is attached to certain logstreams, but the warnings library is like a blunt tool there: you can connect it to your root logstream by `logging.captureWarnings(True)` but that gives you not the control you should have in a production setting.

Reply:
yeah agreed, it's one of the things on the "we should do that"-list. I'll see if I can quickly change that